### PR TITLE
Feature/new tests

### PR DESCRIPTION
**Pull request by lorenzoPrimi**
Not ready to merge yet.

### CODE REVIEW

File: ooniprobe/Model/NetworkMeasurement.mm [hunk 1]
```diff
@@ -464,3 +464,221 @@ -(void)on_entry:(const char*)str{
...
+    test.start([self]() {
+        [self testEnded];
+    });
+}
+
+/*
+ whatsapp: red if "whatsapp_endpoints_status" or "whatsapp_web_status" or "registration_server" are "blocked"
+ docs: https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-018-whatsapp.md#semantics
+ */
+
```
Comment (lorenzoPrimi) [inline ooniprobe/Model/NetworkMeasurement.mm]: @joelanders

File: ooniprobe/Model/NetworkMeasurement.mm [hunk 2]
```diff
@@ -464,3 +464,221 @@ -(void)on_entry:(const char*)str{
...
+    test.start([self]() {
+        [self testEnded];
+    });
+}
+/*
+ for telegram: red if either "telegram_http_blocking" or "telegram_tcp_blocking" is true, OR if ""telegram_web_status" is "blocked"
+ the "*_failure" keys for telegram and whatsapp might indicate a test failure / anomaly
+ docs: https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-020-telegram.md#semantics
+ */
+-(void)on_entry:(const char*)str{
```
Comment (lorenzoPrimi) [inline ooniprobe/Model/NetworkMeasurement.mm]: @joelanders

File: ooniprobe/Model/NetworkMeasurement.mm [hunk 3]
```diff
@@ -464,3 +464,221 @@ -(void)on_entry:(const char*)str{
...
+                if ([json objectForKey:key] == [NSNull null]) {
+                    if (blocking < ANOMALY_ORANGE)
+                        blocking = ANOMALY_ORANGE;
+                }
+                else if ([[json objectForKey:key] boolValue]) {
+                    blocking = ANOMALY_RED;
+                }
+            }
+        }
+        //TODO don't know if needed, maybe remove
```
Comment (lorenzoPrimi) [inline ooniprobe/Model/NetworkMeasurement.mm]: @joelanders review these two TODOS

File: ooniprobe/Model/NetworkMeasurement.mm [hunk 4]
```diff
@@ -464,3 +464,221 @@ -(void)on_entry:(const char*)str{
...
+    test.start([self]() {
+        [self testEnded];
+    });
+}
+
+/*
+ FB: red blocking if either "facebook_tcp_blocking" or "facebook_dns_blocking" is true
+ docs: https://github.com/TheTorProject/ooni-spec/blob/master/test-specs/ts-019-facebook-messenger.md#semantics
+ */
+-(void)on_entry:(const char*)str{
```
Comment (lorenzoPrimi) [inline ooniprobe/Model/NetworkMeasurement.mm]: @joelanders

File: ooniprobe/Model/NetworkMeasurement.mm (hunk 1 above)
Comment (joelanders) [reply ooniprobe/Model/NetworkMeasurement.mm]: LGTM

File: ooniprobe/Model/NetworkMeasurement.mm (hunk 2 above)
Comment (joelanders) [reply ooniprobe/Model/NetworkMeasurement.mm]: LGTM

File: ooniprobe/Model/NetworkMeasurement.mm (hunk 3 above)
Comment (joelanders) [reply ooniprobe/Model/NetworkMeasurement.mm]: I think I would ignore "telegram_web_failure", but keep what you have around "telegram_web_status".

File: ooniprobe/Model/NetworkMeasurement.mm (hunk 4 above)
Comment (joelanders) [reply ooniprobe/Model/NetworkMeasurement.mm]: LGTM

File: ooniprobe/Model/NetworkMeasurement.mm [hunk 5]
```diff
@@ -464,3 +464,221 @@ -(void)on_entry:(const char*)str{
...
+                    if (blocking < ANOMALY_ORANGE)
+                        blocking = ANOMALY_ORANGE;
+                }
+                else if ([[json objectForKey:key] isEqualToString:@"blocked"]) {
+                    blocking = ANOMALY_RED;
+                }
+            }
+        }
+        
+        //TODO don't know if needed, maybe remove
```
Comment (joelanders) [inline ooniprobe/Model/NetworkMeasurement.mm]: Also here, I would ignore "whatsapp_web_failure" and just look at "whatsapp_web_status".

File: ooniprobe/Model/NetworkMeasurement.mm [hunk 6]
```diff
@@ -464,3 +464,210 @@ -(void)on_entry:(const char*)str{
...
+                if ([json objectForKey:key] == [NSNull null]) {
+                    if (blocking < ANOMALY_ORANGE)
+                        blocking = ANOMALY_ORANGE;
+                }
+                else if ([[json objectForKey:key] boolValue]) {
+                    blocking = ANOMALY_RED;
+                }
+            }
+        }
+        if ([json objectForKey:@"telegram_web_status"]{
```
Comment (hellais) [inline ooniprobe/Model/NetworkMeasurement.mm]: As noticed by travis, there is a typo here (you did not close the parens).

**Comment by hellais** [comment]
Link to the line in the CI that shows some typo in the source: https://travis-ci.org/TheTorProject/ooniprobe-ios/builds/305959580#L3515

**Comment by lorenzoPrimi** [comment]
Ready to merge, testflight is being released.
Need some testing on device and simulator

**Pull request merged**